### ISSUE TITLE
flutter_service_worker.js and main.dart.js being fetched from wrong location

### ISSUE THREAD

**Issue by uldall** [flutter/flutter#153032]
### Steps to reproduce

When using Flutter 3.22.3 I can host a Flutter web application using the following commands and access it at `http://localhost:8080/test/`:
```
cup@cup-work2:/media/cup/data/test$ ~/fvm/versions/3.22.3/bin/flutter --version
Flutter 3.22.3 • channel stable • https://github.com/flutter/flutter.git
Framework • revision b0850beeb2 (3 weeks ago) • 2024-07-16 21:43:41 -0700
Engine • revision 235db911ba
Tools • Dart 3.4.4 • DevTools 2.34.3
cup@cup-work2:/media/cup/data/test$ ~/fvm/versions/3.22.3/bin/flutter create bug_3_22_3
Creating project bug_3_22_3...
Resolving dependencies in `bug_3_22_3`... 
Downloading packages... 
Got dependencies in `bug_3_22_3`.
Wrote 129 files.

All done!
You can find general documentation for Flutter at: https://docs.flutter.dev/
Detailed API documentation is available at: https://api.flutter.dev/
If you prefer video documentation, consider: https://www.youtube.com/c/flutterdev

In order to run your application, type:

  $ cd bug_3_22_3
  $ flutter run

Your application code is in bug_3_22_3/lib/main.dart.

cup@cup-work2:/media/cup/data/test$ cd bug_3_22_3
cup@cup-work2:/media/cup/data/test/bug_3_22_3$ ~/fvm/versions/3.22.3/bin/flutter build web

Font asset "CupertinoIcons.ttf" was tree-shaken, reducing it from 257628 to 1172 bytes (99.5% reduction). Tree-shaking can be disabled by providing the --no-tree-shake-icons flag when
building your app.
Font asset "MaterialIcons-Regular.otf" was tree-shaken, reducing it from 1645184 to 7760 bytes (99.5% reduction). Tree-shaking can be disabled by providing the --no-tree-shake-icons
flag when building your app.
Compiling lib/main.dart for the Web...                             22.6s
✓ Built build/web
cup@cup-work2:/media/cup/data/test/bug_3_22_3$ http-server-spa build/web/ index.html 8080
----------------------------------------------
[OK] Serving static files from ./build/web/
[OK] Using the fallback file index.html
[OK] Listening on http://localhost:8080
----------------------------------------------
[OK] GET /test/
[OK] GET /flutter_bootstrap.js
[OK] GET /favicon.png
[OK] GET /manifest.json
[ER] GET /flutter.js.map
[OK] GET /flutter_service_worker.js
[OK] GET /icons/Icon-192.png
[OK] GET /main.dart.js
[OK] GET /index.html
[OK] GET /flutter_bootstrap.js
[OK] GET /assets/AssetManifest.bin.json
[OK] GET /assets/FontManifest.json
[OK] GET /flutter_service_worker.js

```

If I build the same application with Flutter 3.24.0 the flutter_service_worker.js and main.dart.js files are fetched from the wrong location:
```
cup@cup-work2:/media/cup/data/test$ flutter --version
Flutter 3.24.0 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 80c2e84975 (8 days ago) • 2024-07-30 23:06:49 +0700
Engine • revision b8800d88be
Tools • Dart 3.5.0 • DevTools 2.37.2
cup@cup-work2:/media/cup/data/test$ flutter create bug_3_24_0
Creating project bug_3_24_0...
Resolving dependencies in `bug_3_24_0`... 
Downloading packages... 
Got dependencies in `bug_3_24_0`.
Wrote 129 files.

All done!
You can find general documentation for Flutter at: https://docs.flutter.dev/
Detailed API documentation is available at: https://api.flutter.dev/
If you prefer video documentation, consider: https://www.youtube.com/c/flutterdev

In order to run your application, type:

  $ cd bug_3_24_0
  $ flutter run

Your application code is in bug_3_24_0/lib/main.dart.

cup@cup-work2:/media/cup/data/test$ cd bug_3_24_0
cup@cup-work2:/media/cup/data/test/bug_3_24_0$ flutter build web
Resolving dependencies... 
Downloading packages... 
  collection 1.18.0 (1.19.0 available)
  material_color_utilities 0.11.1 (0.12.0 available)
  string_scanner 1.2.0 (1.3.0 available)
  test_api 0.7.2 (0.7.3 available)
Got dependencies!
4 packages have newer versions incompatible with dependency constraints.
Try `flutter pub outdated` for more information.

Font asset "CupertinoIcons.ttf" was tree-shaken, reducing it from 257628 to 1172 bytes (99.5% reduction). Tree-shaking can be disabled by providing the --no-tree-shake-icons flag when
building your app.
Font asset "MaterialIcons-Regular.otf" was tree-shaken, reducing it from 1645184 to 7760 bytes (99.5% reduction). Tree-shaking can be disabled by providing the --no-tree-shake-icons
flag when building your app.
Compiling lib/main.dart for the Web...                             22.5s
✓ Built build/web
cup@cup-work2:/media/cup/data/test/bug_3_24_0$ http-server-spa build/web/ index.html 8080
----------------------------------------------
[OK] Serving static files from ./build/web/
[OK] Using the fallback file index.html
[OK] Listening on http://localhost:8080
----------------------------------------------
[OK] GET /test/
[OK] GET /flutter_bootstrap.js
[OK] GET /manifest.json
[OK] GET /favicon.png
[OK] GET /icons/Icon-192.png
[ER] GET /flutter.js.map
[ER] GET /test/flutter_service_worker.js
[ER] GET /test/main.dart.js

```

Note the errors:
```
[ER] GET /flutter.js.map
[ER] GET /test/flutter_service_worker.js
[ER] GET /test/main.dart.js
```

I am not sure what the `[ER] GET /flutter.js.map` is about. That file is not present in the `build/web/` folder.

### Expected results

Flutter web application can be accessed using a URL path that contains a sub folder.

### Actual results

Flutter web application can not be accessed using a URL path that contains a sub folder.

### Code sample

The Flutter application can be created using `flutter create bug`.

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

Flutter 3.22.3:
![flutter_3_22_3](https://github.com/user-attachments/assets/01fb1594-22f8-4dcb-93e5-e434e41a7e63)

Flutter 3.24.0:
![flutter_3_24_0](https://github.com/user-attachments/assets/c2abb3cd-aa31-42cd-bce3-29f8773961c3)

</details>

### Logs

See output from description.

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
$ flutter doctor -v
[✓] Flutter (Channel stable, 3.24.0, on Ubuntu 23.04 6.2.0-39-generic, locale en_US.UTF-8)
    • Flutter version 3.24.0 on channel stable at /home/cup/snap/flutter/common/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 80c2e84975 (8 days ago), 2024-07-30 23:06:49 +0700
    • Engine revision b8800d88be
    • Dart version 3.5.0
    • DevTools version 2.37.2

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /home/cup/Android/Sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /home/cup/apps/android-studio-2022.3.1.18-linux/android-studio/jbr/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • clang version 10.0.0-4ubuntu1
    • cmake version 3.16.3
    • ninja version 1.10.0
    • pkg-config version 0.29.1

[✓] Android Studio (version 2022.3)
    • Android Studio at /home/cup/apps/android-studio-2022.3.1.18-linux/android-studio
    • Flutter plugin version 75.1.2
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] IntelliJ IDEA Community Edition (version 2024.1)
    • IntelliJ at /snap/intellij-idea-community/515
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.91.1)
    • VS Code at /snap/code/current/usr/share/code
    • Flutter extension version 3.94.0

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Ubuntu 23.04 6.2.0-39-generic
    • Chrome (web)    • chrome • web-javascript • Google Chrome 127.0.6533.88

[✓] Network resources
    • All expected network resources are available.

• No issues found!

```

</details>

**Comment by huycozy** [comment]
Hi @uldall 
I installed [http-server-spa](https://www.npmjs.com/package/http-server-spa) and use it to host the web but couldn't see the issue:
 
<details>
<summary>Demo</summary>


![Screenshot 2024-08-08 at 13 34 33](https://github.com/user-attachments/assets/35cf26a8-1888-475c-bdca-4365914d0eb6)

```console
➜  repro_stable http-server-spa build/web/ index.html 8080
----------------------------------------------
[OK] Serving static files from ./build/web/
[OK] Using the fallback file index.html
[OK] Listening on http://localhost:8080
----------------------------------------------
[OK] GET /
[OK] GET /flutter_bootstrap.js
[OK] GET /favicon.png
[OK] GET /manifest.json
[OK] GET /icons/Icon-192.png
[OK] GET /flutter_service_worker.js
[OK] GET /main.dart.js
[OK] GET /index.html
[OK] GET /flutter_bootstrap.js
[OK] GET /assets/AssetManifest.bin.json
[OK] GET /assets/FontManifest.json
[OK] GET /assets/fonts/MaterialIcons-Regular.otf
[OK] GET /assets/packages/cupertino_icons/assets/CupertinoIcons.ttf
[OK] GET /manifest.json
[OK] GET /icons/Icon-192.png
[OK] GET /flutter_service_worker.js
[OK] GET /test
[OK] GET /favicon.png
[OK] GET /flutter_service_worker.js
```

</details>

Does the issue only happen with `/test/` route? Does it happen on root `/`? 

Can you retry hosting web with `python` or `dhttpd` instead?

**Comment by uldall** [comment]
Hi @huycozy, thank you for looking into this. I had a look at your test output, and the issue with your test is that you are accessing http://localhost:8080/test instead of http://localhost:8080/test/.

If you access http://localhost:8080/test/ it should fail as described.

**Comment by huycozy** [comment]
I see. I can reproduce it on stable 3.24.0 now. But it seems to be fixed on master channel (`3.24.0-1.0.pre.508`). Can you switch to master and confirm?

```
flutter channel master
flutter upgrade
```

**Comment by uldall** [comment]
I can confirm that the bug is no longer present on master :+1: :
```
cup@cup-work2:/media/cup/data/test/bug_3_24_0$ flutter --version
Flutter 3.24.0-1.0.pre.509 • channel master • https://github.com/flutter/flutter.git
Framework • revision 925ad4e6fa (6 hours ago) • 2024-08-08 00:38:29 -0400
Engine • revision 94cf8c8fad
Tools • Dart 3.6.0 (build 3.6.0-127.0.dev) • DevTools 2.38.0
cup@cup-work2:/media/cup/data/test/bug_3_24_0$ http-server-spa build/web/ index.html 8080
----------------------------------------------
[OK] Serving static files from ./build/web/
[OK] Using the fallback file index.html
[OK] Listening on http://localhost:8080
----------------------------------------------
[OK] GET /test/
[OK] GET /flutter_bootstrap.js
[OK] GET /manifest.json
[OK] GET /favicon.png
[OK] GET /flutter_service_worker.js
[OK] GET /icons/Icon-192.png
[OK] GET /main.dart.js
[OK] GET /index.html
[OK] GET /flutter_bootstrap.js
[OK] GET /assets/AssetManifest.bin.json
[OK] GET /assets/FontManifest.json
[OK] GET /assets/fonts/MaterialIcons-Regular.otf
[OK] GET /assets/packages/cupertino_icons/assets/CupertinoIcons.ttf

```
